fourier/rfft.go

```diff
@@ -0,0 +1,1151 @@
+//
+//  ifac   A work array containing the factors of n. ifac must have
+//         length of at least 15.
+//
+//  Output parameters:
+//
+//  r      r[0] = the sum from i=0 to i=n-1 of r[i]
+//
+//         if n is even set l=n/2, if n is odd set l = (n+1)/2
+//           then for k = 1, ..., l-1
+//             r[2*k-1] = the sum from i = 0 to i = n-1 of
+//               r[i]*cos(k*i*2*pi/n)
+//             r[2*k] = the sum from i = 0 to i = n-1 of
+//               -r[i]*sin(k*i*2*pi/n)
+//
+//         if n is even
+//           r[n-1] = the sum from i = 0 to i = n-1 of
+//             (-1)^i*r[i]
+//
+//  This transform is unnormalized since a call of rfftf
+//  followed by a call of rfftb will multiply the input
+//  sequence by n.
+//
+//  work   contains results which must not be destroyed between
+//         calls of rfftf or rfftb.
+//  ifac   contains results which must not be destroyed between
+//         calls of rfftf or rfftb.
+func rfftf(n int, r, work []float64, ifac []int) {
+	if len(r) < n {
+		panic("fourier: short sequence")
+	}
+	if len(work) < 2*n {
+		panic("fourier: short work")
+	}
+	if len(ifac) < 15 {
+		panic("fourier: short ifac")
+	}
+	if n == 1 {
+		return
+	}
+	rfftf1(n, r[:n], work[:n], work[n:2*n], ifac[:15])
+}
+
+func rfftf1(n int, c, ch, wa []float64, ifac []int) {
+	nf := ifac[1]
+	na := true
+	l2 := n
+	iw := n - 1
+
+	for k1 := 1; k1 <= nf; k1++ {
+		kh := nf - k1
+		ip := ifac[kh+2]
+		l1 := l2 / ip
+		ido := n / l2
+		idl1 := ido * l1
+		iw -= (ip - 1) * ido
+		na = !na
+
+		switch ip {
+		case 4:
+			ix2 := iw + ido
+			ix3 := ix2 + ido
+			if na {
+				radf4(ido, l1, ch, c, wa[iw:], wa[ix2:], wa[ix3:])
+			} else {
+				radf4(ido, l1, c, ch, wa[iw:], wa[ix2:], wa[ix3:])
+			}
+		case 2:
+			if na {
+				radf2(ido, l1, ch, c, wa[iw:])
+			} else {
+				radf2(ido, l1, c, ch, wa[iw:])
+			}
+		case 3:
+			ix2 := iw + ido
+			if na {
+				radf3(ido, l1, ch, c, wa[iw:], wa[ix2:])
+			} else {
+				radf3(ido, l1, c, ch, wa[iw:], wa[ix2:])
+			}
+		case 5:
+			ix2 := iw + ido
+			ix3 := ix2 + ido
+			ix4 := ix3 + ido
+			if na {
+				radf5(ido, l1, ch, c, wa[iw:], wa[ix2:], wa[ix3:], wa[ix4:])
+			} else {
+				radf5(ido, l1, c, ch, wa[iw:], wa[ix2:], wa[ix3:], wa[ix4:])
+			}
+		default:
+			if ido == 1 {
+				na = !na
+			}
+			if na {
+				radfg(ido, ip, l1, idl1, ch, ch, ch, c, c, wa[iw:])
+				na = false
+			} else {
+				radfg(ido, ip, l1, idl1, c, c, c, ch, ch, wa[iw:])
+				na = true
+			}
+		}
+
+		l2 = l1
+	}
+
+	if na {
+		return
+	}
+	for i := 0; i < n; i++ {
+		c[i] = ch[i]
+	}
+}
+
+func radf2(ido, l1 int, cc, ch, wa1 []float64) {
+	cc3 := newThreeArray(ido, l1, 2, cc)
+	ch3 := newThreeArray(ido, 2, l1, ch)
+
+	for k := 0; k < l1; k++ {
+		ch3.set(0, 0, k, cc3.at(0, k, 0)+cc3.at(0, k, 1))
+		ch3.set(ido-1, 1, k, cc3.at(0, k, 0)-cc3.at(0, k, 1))
+	}
+	if ido < 2 {
+		return
+	}
+	if ido > 2 {
+		idp2 := ido + 1
+		for k := 0; k < l1; k++ {
+			for i := 2; i < ido; i += 2 {
```

Comment:
Shouldn't this be `i <= ido`? Looks like `idp2` is zero-indexed (only `ido + 1`)

Reply:
That would suggest all of the loops with an iteration bound of `ido` would be converted to `<=`. That change results in a panic.

Reply:
`idp2` is badly named, but the name is retained to help track the code.

Reply:
I should probably explain the process I used to make the transformation from 1-based to 0-based. For each function in turn I swapped from the 1-based types to the 0-based types, fixing up the indexing at the call site by adding `+1` to the indexing parameters. I then algebraically smoothed out the change, simplifying as much as I could. Also note that `ido` is defined in the outer functions as `n / l2`. These functions are not modified in either the trivial or subtle commits.